lib/awful/keygrabber.lua

```diff
@@ -45,7 +207,411 @@ function keygrabber.stop(g)
+
+    -- Ease making keygrabber that wont hang forever if no action is taken.
+    if self.timeout and not self._private.timer then
+        self._private.timer = gtimer {
+            timeout     = self.timeout,
+            single_shot = true,
+            callback    = function()
+                if self.timeout_callback then
+                    self.timeout_callback(self)
+                end
+                self:stop()
+            end
+        }
+    end
+
+    if self._private.timer then
+        self._private.timer:start()
+        assert(self._private.timer.data.source_id)
+    end
+
+    keygrab.current_instance = self
+
+    keygrab.emit_signal("property::current_instance", keygrab.current_instance)
+
+    self:emit_signal("started")
+end
+
+--- Stop the keygrabber.
+-- @function keygrabber:stop
+function keygrabber:stop(_release_key, _release_mods) -- (at)function disables ldoc params
+    keygrab.stop(self.grabber)
+
+    if self.stop_callback then
+        self.stop_callback(
+            self.current_instance, _release_key, _release_mods, self.sequence
+        )
+    end
+
+    self.current_instance = nil
+
+    keygrab.current_instance = nil
+    keygrab.emit_signal("property::current_instance", nil)
+
+    self.grabber = nil
+    self:emit_signal("stopped")
+end
+
+--- Add a keybindings to this keygrabber.
+--
+-- Those keybindings will automatically start the keygrabbing when hit.
+--
+-- @tparam table mods A table with modifier keys, such as `shift`, `mod4`, `mod1` (alt) or
+--  `control`.
+-- @tparam string key The key name, such as `left` or `f`
+-- @tparam function callback A function that will be called when the key
+-- combination is pressed.
+-- @tparam[opt] table description A table various metadata to be used for `awful.hotkeys_popup`.
+-- @tparam string description.description The keybinding description
+-- @tparam string description.group The keybinding group
+function keygrabber:add_keybinding(mods, key, callback, description)
+    self._private.keybindings[key] = self._private.keybindings[key] or {}
+    self._private.keybindings[key][#self._private.keybindings[key]+1] = {
+        mods, key, callback, description
+    }
+
+    if self.export_keybindings then
+        add_root_keybindings(self, {{mods, key, callback, description}})
+    end
+end
+
+function keygrabber:set_root_keybindings(keys)
+    add_root_keybindings(self, keys)
+end
+
+-- Turn into a set.
+function keygrabber:set_allowed_keys(keys)
+    self._private.allowed_keys = {}
+
+    for _, v in ipairs(keys) do
+        self._private.allowed_keys[v] = true
+    end
+end
+
+--- When the keygrabber starts.
+-- @signal started
+
+--- When the keygrabber stops.
+-- @signal stopped
+
+--- A function called when a keygrabber starts.
+-- @callback start_callback
+-- @tparam keygrabber keygrabber The keygrabber.
+
+--- The callback function to always call without arguments, regardless of
+-- whether the prompt was cancelled.
```

Comment:
Huh? Copy&paste from `awful.prompt`, or am I missing something?
Also, you document this function as `call[ed] without arguments` and then describe the arguments?